ARROW-14671: [Python][Doc] Documentation on how to integrate PyArrow and R (#11679)
apache/arrow: docs/source/python/integration/python_r.rst

    # Create the pyarrow array we want to pass to R
    import pyarrow
    array = pyarrow.array((1, 2, 3))

    # Import the pyarrow module that provides access to the C Data interface
    from pyarrow.cffi import ffi as arrow_c

    # Allocate structures where we will export the Array data 
    # and the Array schema. They will be released when we exit the with block.
    with arrow_c.new("struct ArrowArray*") as c_array, \
         arrow_c.new("struct ArrowSchema*") as c_schema:
        # Get the references to the C Data structures.
        c_array_ptr = int(arrow_c.cast("uintptr_t", c_array))
        c_schema_ptr = int(arrow_c.cast("uintptr_t", c_schema))

        # Export the Array and its schema to the C Data structures.
        array._export_to_c(c_array_ptr)
        array.type._export_to_c(c_schema_ptr)

        # Invoke the R addthree_cdata function passing the references
        # to the array and schema C Data structures. 
        # Those references are passed as strings as R doesn't have
        # native support for 64bit integers, so the integers are
        # converted to their string representation for R to convert it back.

>         r_result_array = addthree(str(c_array_ptr), str(c_schema_ptr))

Reviewer: ```suggestion
        r_result_array = addthree_cdata(str(c_array_ptr), str(c_schema_ptr))
```

.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

Integrating PyArrow with R
==========================

Arrow supports exchanging data within the same process through the
:ref:`c-data-interface`.

This can be used to exchange data between Python and R functions and
methods so that the two languages can interact without any cost of
marshaling and unmarshaling data.

.. note::

    The article takes for granted that you have a ``Python`` environment
    with ``pyarrow`` correctly installed and an ``R`` environment with
    ``arrow`` library correctly installed. 
    See `Python Install Instructions <https://arrow.apache.org/docs/python/install.html>`_
    and `R Install instructions <https://arrow.apache.org/docs/r/#installation>`_
    for further details.

Invoking R functions from Python
--------------------------------

Suppose we have a simple R function receiving an Arrow Array to
add ``3`` to all its elements:

.. code-block:: R

    library(arrow)

    addthree <- function(arr) {
        return(arr + 3L)
    }

> We could save such function in a ``addthree.R`` file so that we can

Reviewer 2: ```suggestion
We could save such a function in a ``addthree.R`` file so that we can
```

make it available for reuse.

> Once the ``addthree.R`` is created we can invoke any of its functions

Reviewer 2: ```suggestion
Once the ``addthree.R`` file is created we can invoke any of its functions
```

``rpy2`` can be installed using ``pip`` like most python libraries

.. code-block:: bash

    $ pip install rpy2

The most basic thing we can do with our ``addthree`` function is to
invoke it from Python with a number and see how it will return the result.

To do so we can create an ``addthree.py`` file which uses ``rpy2`` to
import the ``addthree`` function from ``addthree.R`` file and invoke it:

.. code-block:: python

    import rpy2.robjects as robjects

    # Load the addthree.R file
    r_source = robjects.r["source"]
    r_source("addthree.R")

    # Get a reference to the addthree function
    addthree = robjects.r["addthree"]

    # Invoke the function
    r = addthree(3)

    # Access the returned value
    value = r[0]
    print(value)

Running the ``addthree.py`` file will show how our Python code is able
to access the ``R`` function and print the expected result:

.. code-block:: bash

    $ python addthree.py 
    6.0

If instead of passing around basic data types we want to pass around

> Arrow Arrays, we can do so relying on the ``rpy2-arrow`` module which

Reviewer 2: Give a link to the `rpy2-arrow` homepage?

Author: :+1:

implements rpy2 support for Arrow types.

> ``rpy2`` can be installed through ``pip``:

Reviewer 2: You probably mean "rpy2-arrow".

Author: :+1: fixed

.. code-block:: bash

    $ pip install rpy2-arrow

> ``rpy2-arrow`` implements converters from pyarrow objects to r arrow objects,

Reviewer 2: Capitalization: "PyArrow", "R", "Arrow". `arrow` and `pyarrow` should only be used when referring to the actual C++ (resp. Python) namespace.

Author: :+1:

    $ pip install rpy2-arrow

``rpy2-arrow`` implements converters from PyArrow objects to R Arrow objects,

> this is done without occurring into any data copy cost as it relies on the

Reviewer 2: ```suggestion
this is done without incurring any data copy cost as it relies on the
```

    # using the rarrow_to_py_array function
    py_result = rarrow_to_py_array(r_result)
    print("RESULT", type(py_result), py_result)

Running the newly modified ``addthree.py`` should now properly execute
the R function and print the resulting pyarrow Array:

.. code-block:: bash

    $ python addthree.py
    RESULT <class 'pyarrow.lib.DoubleArray'> [
      4,
      5,
      6
    ]

.. note::

    Even though we sent an ``Int64Array`` to R, we end up with a
    result as a ``DoubleArray``. That's due to the lack of native
    support for 64 bits numbers in R and thus its use of doubles
    to represent those.

For additional information you can refer to
`rpy2 Documentation <https://rpy2.github.io/doc/latest/html/index.html>`_
and `rpy2-arrow Documentation <https://rpy2.github.io/rpy2-arrow/version/main/html/index.html>`_

Invoking Python functions from R
--------------------------------

Exposing Python functions to R can be done through the ``reticulate``
library. For example if we want to invoke :func:`pyarrow.compute.add` from
R on an Array created in R we can do so importing ``pyarrow`` in R
through ``reticulate``.

> A basic ``addthree.R`` script that invokes ``add`` to add ``3`` to

Reviewer 2: Do we want to expand on this here or should we just point to the existing docs [on the R side](https://arrow.apache.org/docs/r/articles/python.html#using)?

cc @thisisnic  and @jonkeane

Reviewer 3: I'm fine duplicating a bit in both places, but we should link as well

Author: Added link at the end of the paragraph near to the Reticulate one

    # Load arrow and reticulate libraries
    library(arrow)
    library(reticulate)

    # Create a new array in R
    a <- Array$create(c(1, 2, 3))

    # Make pyarrow.compute available to R
    pc <- import("pyarrow.compute")

    # Invoke pyarrow.compute.add with the array and 3
    # This will add 3 to all elements of the array and return a new Array
    result <- pc$add(a, 3)

    # Print the result to confirm it's what we expect
    print(result)

Invoking the ``addthree.R`` script will print the outcome of adding
``3`` to all the elements of the original ``Array$create(c(1, 2, 3))`` array:

.. code-block:: bash

    $ R --silent -f addthree.R 
    Array
    <double>
    [
      4,
      5,
      6
    ]

For additional information you can refer to
`Reticulate Documentation <https://rstudio.github.io/reticulate/>`_
and to the `R Arrow documentation <https://arrow.apache.org/docs/r/articles/python.html#using>`_

> R to Python communication using C Data Interface
> ------------------------------------------------

Reviewer 2: ```suggestion
R to Python communication using the C Data Interface
----------------------------------------------------
```

> Both the solutions described in previous chapters use the Arrow C Data

Reviewer 2: ```suggestion
Both solutions described above use the Arrow C Data
```

interface under the hood.

In case we want to extend the previous ``addthree`` example to switch
from using ``rpy2-arrow`` to using the plain C Data interface we can
do so by introducing some modifications to our codebase.

To enable importing the Arrow Array from the C Data interface we have to
wrap our ``addthree`` function in a function that does the extra work
necessary to import an Arrow Array in R from the C Data interface.

That work will be done by the ``addthree_cdata`` function which invokes the

> ``addthree`` function once the Array is imported.

Reviewer 2: This entire snippet doesn't seem very informative to me ("introduce some modifications"... "do the extra work"...). Perhaps condense them into one or two meaningful sentences?

Also, this does not explain _why_ I would want to avoid ``rpy2-arrow`` and instead invoke the C Data Interface directly.

Reviewer 4: @amol- please address this in a follow-up.

Our ``addthree.R`` will thus have both the ``addthree_cdata`` and the 
``addthree`` functions:

.. code-block:: R

    library(arrow)

    addthree_cdata <- function(array_ptr_s, schema_ptr_s) {
        array_ptr <- as.numeric(array_ptr_s)
        schema_ptr <- as.numeric(schema_ptr_s)

>         a <- Array$import_from_c(array_ptr, schema_ptr)

Reviewer 2: Would it make sense to have `Array$import_from_c` accept both strings and numerics, to avoid defining this function here? Or perhaps we should simply wait for the R pointer improvements to land before making a decision on this PR, anyway?

@paleolimbot

    with arrow_c.new("struct ArrowArray*") as c_array, \
         arrow_c.new("struct ArrowSchema*") as c_schema:
        # Get the references to the C Data structures.
        c_array_ptr = int(arrow_c.cast("uintptr_t", c_array))
        c_schema_ptr = int(arrow_c.cast("uintptr_t", c_schema))

        # Export the Array and its schema to the C Data structures.
        array._export_to_c(c_array_ptr)
        array.type._export_to_c(c_schema_ptr)

        # Invoke the R addthree_cdata function passing the references
        # to the array and schema C Data structures. 
        # Those references are passed as strings as R doesn't have
        # native support for 64bit integers, so the integers are
        # converted to their string representation for R to convert it back.
        r_result_array = addthree_cdata(str(c_array_ptr), str(c_schema_ptr))

        # r_result will be an Environment variable that contains the
        # arrow Array built from R as the return value of addthree.
        # To make it available as a Python pyarrow array we need to export
        # it as a C Data structure invoking the Array$export_to_c R method

>         r_result_array["export_to_c"](float(c_array_ptr), float(c_schema_ptr))

Reviewer 2: Ouch, why does this cast to `float`? Floats cannot represent all 64-bit numbers faithfully.

Author: That's sadly a side effect of interfacing to R. Integers in R are 32-bits and thus if you try to pass around the pointers as integers they will overflow. You get a `integer 140411646274592 does not fit '32-bit int'` kind of error.

For the `addthree_cdata` function (as it was implemented by myself) I passed the pointers as strings, so they would represent the exact value, but for `Array$export_to_c` as it's a R function expecting a number I couldn't provide a string, I had to pass a `REAL` as expected by the function. Thus I had to go for a `float` to have `rpy2` convert it to a `REAL`.

I guess an alternative might be to get a reference to the `as.numeric` R function and invoke `as.numeric(str(c_array_ptr))` that would force the conversion to an R numeric from the string in R itself so we would avoid going through a Python float.

Reviewer 2: What does it change to call `as.numeric`? You would still get a floating-point number with not enough precision, right?

Author: Changed the example to use `as.numeric` instead of `float`

Author: @pitrou Well, I tried with `bit64` `integer64` type, but it didn't work, it seems to only accept a REAL, if I pass a `integer64` to `export_to_c` or `import_from_c` I get a segfault. Someone with more R knowledge might want to chime in. @jonkeane @thisisnic

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

Integrating PyArrow with R
==========================

Arrow supports exchanging data within the same process through the

> :ref:`c-data-interface`.

Reviewer 3: Would it be worth spelling out here that this also works for subprocesses? That might be assumed / known for some with the phrase "within the same process" but I also think others would think / assume that subprocess don't count as the same process for this purpose.

Author: Uhm, that's an interesting point. I _think_ that suggesting it here might open a can of worms as what you have access to will depend a lot on _when_ the subprocesses are forked and people might complain that it doesn't work. I would act safe and only explicitly state about same process leaving to the reader knowledge to imagine it can also work for subprocesses under some circumstances.

Reviewer 2: Agreed with @amol- . Users who know and understand what forking are able to make out the consequences by themselves; users who don't don't deserve to be confused :-)

This can be used to exchange data between Python and R functions and
methods so that the two languages can interact without any cost of
marshaling and unmarshaling data.

.. note::

    The article takes for granted that you have a ``Python`` environment
    with ``pyarrow`` correctly installed and an ``R`` environment with

>     ``arrow`` library correctly installed.

Reviewer 3: Could we? Should we? link to https://arrow.apache.org/docs/python/install.html and https://arrow.apache.org/docs/r/#installation here?

Author: Added link, I guess we could also remove this note. I felt it was useful mostly because if you approach it from a Python developer point of view it might not be immediately obvious that you need a working R environment too. But in the end we don't state on every doc page that to follow the doc you need a working Arrow setup.

Reviewer 3: Thanks for this! I'm happy either way (to leave it in or take it out). It _is_ nice to have the links for building both in case one or the other is less familiar to someone one – but you're absolutely right we don't start each doc with "to build Arrow..."

`rpy2-arrow <https://rpy2.github.io/rpy2-arrow/version/main/html/index.html>`_ 
module which implements ``rpy2`` support for Arrow types.

``rpy2-arrow`` can be installed through ``pip``:

.. code-block:: bash

    $ pip install rpy2-arrow

``rpy2-arrow`` implements converters from PyArrow objects to R Arrow objects,
this is done without occurring into any data copy cost as it relies on the
C Data interface.

To pass to ``addthree`` a PyArrow array our ``addthree.py`` needs to be modified
to enable ``rpy2-arrow`` converters and then pass the PyArrow array:

.. code-block:: python

    import rpy2.robjects as robjects
    from rpy2_arrow.pyarrow_rarrow import (rarrow_to_py_array,
                                           converter as arrowconverter)
    from rpy2.robjects.conversion import localconverter

    r_source = robjects.r["source"]
    r_source("addthree.R")

    addthree = robjects.r["addthree"]

    import pyarrow

    array = pyarrow.array((1, 2, 3))

    # Enable rpy2-arrow converter so that R can receive the array.
    with localconverter(arrowconverter):
        r_result = addthree(array)

    # The result of the R function will be an R Environment

>     # we can convert back the Environment to a pyarrow Array

Reviewer 3: ```suggestion
    # we can convert the Environment back to a pyarrow Array
```

I think this makes it a little bit clearer

    # using the rarrow_to_py_array function
    py_result = rarrow_to_py_array(r_result)
    print("RESULT", type(py_result), py_result)

Running the newly modified ``addthree.py`` should now properly execute
the R function and print the resulting PyArrow Array:

.. code-block:: bash

    $ python addthree.py
    RESULT <class 'pyarrow.lib.DoubleArray'> [
      4,
      5,
      6
    ]

.. note::

    Even though we sent an ``Int64Array`` to R, we end up with a
    result as a ``DoubleArray``. That's due to the lack of native
    support for 64 bits numbers in R and thus its use of doubles

>     to represent those.

Reviewer 3: If the script were instead (note `3L` versus `3`) I _think_ we would get an integer back.

```
    library(arrow)

    addthree <- function(arr) {
        return(arr + 3L)
    }
```

Author: Good catch! I actually forgot that R goes for numerics by default

and `rpy2-arrow Documentation <https://rpy2.github.io/rpy2-arrow/version/main/html/index.html>`_

Invoking Python functions from R
--------------------------------

Exposing Python functions to R can be done through the ``reticulate``
library. For example if we want to invoke :func:`pyarrow.compute.add` from
R on an Array created in R we can do so importing ``pyarrow`` in R
through ``reticulate``.

A basic ``addthree.R`` script that invokes ``add`` to add ``3`` to
an R array would look like:

.. code-block:: R

    # Load arrow and reticulate libraries
    library(arrow)
    library(reticulate)

    # Create a new array in R
    a <- Array$create(c(1, 2, 3))

    # Make pyarrow.compute available to R
    pc <- import("pyarrow.compute")

    # Invoke pyarrow.compute.add with the array and 3
    # This will add 3 to all elements of the array and return a new Array
    result <- pc$add(a, 3)

    # Print the result to confirm it's what we expect
    print(result)

Invoking the ``addthree.R`` script will print the outcome of adding
``3`` to all the elements of the original ``Array$create(c(1, 2, 3))`` array:

.. code-block:: bash

>     $ R --slave -f addthree.R 

Reviewer 3: ```suggestion
    $ R --silent -f addthree.R 
```

Does the same thing, with less of a negative connotation to it
